packages/react-reconciler/src/ReactFiberCommitWork.new.js

```diff
@@ -507,57 +506,55 @@ function commitHookEffectListMount2(fiber: Fiber): void {
   }
 }
 
-export function commitPassiveEffectDurations(
+function commitProfilerPassiveEffect(
   finishedRoot: FiberRoot,
   finishedWork: Fiber,
 ): void {
   if (enableProfilerTimer && enableProfilerCommitHooks) {
-    // Only Profilers with work in their subtree will have an Update effect scheduled.
-    if ((finishedWork.flags & Update) !== NoFlags) {
-      switch (finishedWork.tag) {
-        case Profiler: {
-          const {passiveEffectDuration} = finishedWork.stateNode;
-          const {id, onPostCommit} = finishedWork.memoizedProps;
-
-          // This value will still reflect the previous commit phase.
-          // It does not get reset until the start of the next commit phase.
-          const commitTime = getCommitTime();
-
-          if (typeof onPostCommit === 'function') {
-            if (enableSchedulerTracing) {
-              onPostCommit(
-                id,
-                finishedWork.alternate === null ? 'mount' : 'update',
-                passiveEffectDuration,
-                commitTime,
-                finishedRoot.memoizedInteractions,
-              );
-            } else {
-              onPostCommit(
-                id,
-                finishedWork.alternate === null ? 'mount' : 'update',
-                passiveEffectDuration,
-                commitTime,
-              );
-            }
+    switch (finishedWork.tag) {
+      case Profiler: {
+        const {passiveEffectDuration} = finishedWork.stateNode;
+        const {id, onPostCommit} = finishedWork.memoizedProps;
+
+        // This value will still reflect the previous commit phase.
+        // It does not get reset until the start of the next commit phase.
+        const commitTime = getCommitTime();
+
+        if (typeof onPostCommit === 'function') {
+          if (enableSchedulerTracing) {
+            onPostCommit(
+              id,
+              finishedWork.alternate === null ? 'mount' : 'update',
+              passiveEffectDuration,
+              commitTime,
+              finishedRoot.memoizedInteractions,
+            );
+          } else {
+            onPostCommit(
+              id,
+              finishedWork.alternate === null ? 'mount' : 'update',
+              passiveEffectDuration,
+              commitTime,
+            );
           }
+        }
 
-          // Bubble times to the next nearest ancestor Profiler.
-          // After we process that Profiler, we'll bubble further up.
-          let parentFiber = finishedWork.return;
-          while (parentFiber !== null) {
-            if (parentFiber.tag === Profiler) {
-              const parentStateNode = parentFiber.stateNode;
-              parentStateNode.passiveEffectDuration += passiveEffectDuration;
-              break;
-            }
-            parentFiber = parentFiber.return;
+        // Bubble times to the next nearest ancestor Profiler.
+        // After we process that Profiler, we'll bubble further up.
+        // TODO: Use JS Stack instead
```

Comment:
Roughly this (typed it out fast, might be a mistake)

```js
let nearestProfiler = null;

function commitWork(finishedWork) {
  switch (finishedWork.tag) {
    case Profiler: {
      // Can push/pop values onto stack before entering/leaving a subtree, like
      // how we do in the render phase, except since this is a recursive
      // function we can use the regular JS stack instead of our fake
      // fiber stack.
      //
      // This is the code for Profiler. We'd do something similar for Offscreen.
      if (finishedWork.subtreeFlags & FlagForThisPhase) {
        const prevProfilerOnStack = nearestProfiler;
        nearestProfiler = finishedWork;

        try {
          let child = finishedWork.child;
          while (child !== null) {
            commitWork(child);
            child = child.sibling;
          }
        } finally {
          nearestProfiler = prevProfilerOnStack;
        }
      }

      commitProfilerEffect(finishedWork);
      break;
    }
    default: {
      // Normal path, no pushing or popping
      if (finishedWork.subtreeFlags & FlagForThisPhase) {
        let child = finishedWork.child;
        while (child !== null) {
          commitWork(child);
          child = child.sibling;
        }
      }

      switch (finishedWork.tag) {
        case FunctionComponent: {
          // Can reference `nearestProfiler` here
          commitFunctionComponent(finishedWork, nearestProfiler);
          break;
        }
        // ... and so on
      }
      break;
    }
  }
}

```

Reply:
Happy to pick this TODO up if you haven't already started on it. Let me know.

Reply:
Circling back on this.

It's possible that I'm misunderstanding what you're suggesting, but I think the pseudo-code solution is missing a few things.

There are two places we walk the return path to find the nearest Profiler ancestor:
* `commitLifeCycles` to bubble up layout effect durations from one Profiler to the next.
* `commitProfilerPassiveEffect` to bubble up passive effect durations from one Profiler to the next.

Neither of these functions are recursive. The work loop functions that call them are (e.g. `commitLayoutEffects` and `flushPassiveMountEffects`) so maybe that's what you meant to show instead with the above pseudo-code example?

There's another snag though. These durations bubble up one Profiler at a time. The "TODO" code in question is just bubbling the durations from the current Profiler to the next nearest (ancestor) Profiler. If we used a stack for this, the top of the stack would always just point to the current Profiler (which we already have, since its the current Fiber we're working on). So we'd need to maintain pointers to the *top two* Profilers in the stack.

This could work, but it's not as clean an implementation as you were perhaps thinking? I'll post a PR.

Reply:
#19895

Reply:
> The work loop functions that call them are (e.g. commitLayoutEffects and flushPassiveMountEffects) so maybe that's what you meant to show instead with the above pseudo-code example?

Yeah `commitWork` in my example is meant to correspond to `commitLayoutEffects` et al

Reply:
Gotcha.

The naming of `commitWork` confused the heck out of me then, given where the TODO comments were previously (in `ReactFiberCommitWork` - which also has a function called `commitWork`)